NeoMathEngine/src/CPU/x86/avx/CpuMathEngineDnnConvAvx.cpp

```diff
@@ -0,0 +1,529 @@
+/* Copyright © 2017-2020 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <cmath>
+
+#include <immintrin.h>
+#include <NeoMathEngine/NeoMathEngine.h>
+
+#include <CpuMathEngine.h>
+#include <MathEngineDnnConv.h>
+#include <MathEngineCommon.h>
+#include <MemoryHandleInternal.h>
+
+namespace NeoML {
+
+template<int imm8, int FC>
+inline void Process_avx( __m256& r0, __m256& r1, __m256& r2,
+						 __m128& s0, const float* &fltPtr ) {
+	__m128 st0_m128 =  _mm_permute_ps( s0, imm8 );
+	__m256 st0 = _mm256_set_m128( st0_m128, st0_m128 );
+
+	__m256 f0 = _mm256_loadu_ps( fltPtr );
+	__m256 f1 = _mm256_loadu_ps( fltPtr + 8 );
+	__m256 f2 = _mm256_loadu_ps( fltPtr + 16 );
+
+	r0 = _mm256_fmadd_ps( f0, st0, r0 );
+	r1 = _mm256_fmadd_ps( f1, st0, r1 );
+	r2 = _mm256_fmadd_ps( f2, st0, r2 );
+
+	fltPtr += FC;
+}
+
+template<int imm8, int FC>
+inline void Process_avx_x2( __m256& r00, __m256& r01, __m256& r02,
+							__m256& r10, __m256& r11, __m256& r12,
+							__m128& s0, __m128& s1, const float* &fltPtr ) {
+	__m128 st0_m128 =  _mm_permute_ps( s0, imm8 );
+	__m256 st0 = _mm256_set_m128( st0_m128, st0_m128 );
+	__m128 st1_m128 =  _mm_permute_ps( s1, imm8 );
+	__m256 st1 = _mm256_set_m128( st1_m128, st1_m128 );
+
+	__m256 f0 = _mm256_loadu_ps( fltPtr );
+	__m256 f1 = _mm256_loadu_ps( fltPtr + 8 );
+	__m256 f2 = _mm256_loadu_ps( fltPtr + 16 );
+
+	r00 = _mm256_fmadd_ps( f0, st0, r00 );
+	r01 = _mm256_fmadd_ps( f1, st0, r01 );
+	r02 = _mm256_fmadd_ps( f2, st0, r02 );
+
+	r10 = _mm256_fmadd_ps( f0, st1, r10 );
+	r11 = _mm256_fmadd_ps( f1, st1, r11 );
+	r12 = _mm256_fmadd_ps( f2, st1, r12 );
+
+	fltPtr += FC;
+}
+
+extern "C"
+FME_DLL_EXPORT void BlobConvolution_avx_f3x3_c24_fc24( IMathEngine& mathEngine, int threadCount, const CCommonConvolutionDesc& desc, const float* sourceData,
```

Comment:
this huge function needs refactoring. at least move autos to separate functions

Reply:
lambdas*

Reply:
If I use separate function I have to pass all necessary constants into them. It will impact performance and code clarity not in positive manner.

Reply:
they are passed to lambdas as well by reference (as you mention [&])
moreover, you may inline them

Reply:
Hence I should declare all variables as global. :(

Reply:
why? just enhance arguments list